### PR TITLE
More focused problem matcher regex

### PR DESCRIPTION
**Pull request by bryanmacfarlane**
see: https://github.com/actions/setup-go/issues/46#issuecomment-604937131

The main change in this PR is the regex pattern in `matcher.json`.  However, I added some tests to validate it catches cases.

Instead of creating one general and broad regex to catch everything, our problem matchers should be more focused and if there's cases we miss, the platform supports adding many problem matchers.

setup-go is also about to release `v2` with the current tag being `v2-beta` so this is the perfect time to try something more restrictive.  If we find over time there's other error cases, we will add additional patterns.

### CODE REVIEW

File: __tests__/setup-go.test.ts [hunk 1]
```diff
@@ -9,6 +9,9 @@ import * as main from '../src/main';
 import * as im from '../src/installer';
 
 let goJsonData = require('./data/golang-dl.json');
+let matchers = require('../matchers.json');
+let matcherPattern = matchers.problemMatcher[0].pattern[0];
+let matcherRegExp = new RegExp(matcherPattern.regexp);
```
Comment (ericsciple) [inline __tests__/setup-go.test.ts]: cool

File: matchers.json [hunk 2]
```diff
@@ -4,11 +4,11 @@
             "owner": "go",
             "pattern": [
                 {
-                    "regexp": "^([^:]*: )?((.:)?[^:]*):(\\d+)(:(\\d+))?: (.*)$",
-                    "file": 2,
-                    "line": 4,
-                    "column": 6,
-                    "message": 7
+                    "regexp": "((?:\\.{0,2}\\/|\\.{0,2}\\\\)(?:.+\\.go)):(?:(\\d+):(\\d+):)? (.*)",
```
Comment (ericsciple) [inline matchers.json]: nit: It looks like you might be able to eliminate some groups. I think this is equivalent:

```
"(\\.{0,2}[\\/\\\\].+\\.go):(?:(\\d+):(\\d+):)? (.*)",
```

- moved `\\.{0,2}` in front of the group
- then switched `(?:\\/|\\\\)` to a char set `[\\/\\\\]` - arguably simpler? idk
- Removed group around the rest of the file path `(?:.+\\.go)` =\> `.+\\.go`

File: __tests__/setup-go.test.ts [hunk 3]
```diff
@@ -322,6 +325,100 @@ describe('setup-go', () => {
...
+    let line = '..\\main.go:13:2: undefined: fmt.Printl';
+    let annotation = testMatch(line);
+    expect(annotation).toBeDefined();
+    expect(annotation.line).toBe(13);
+    expect(annotation.column).toBe(2);
+    expect(annotation.file).toBe('..\\main.go');
+    expect(annotation.message).toBe('undefined: fmt.Printl');
+  });
+
+  it('only matches leading dots on unix path', async () => {
```
Comment (ericsciple) [inline __tests__/setup-go.test.ts]: The description doesnt align with the behavior

Should the description be updated, or should the behavior be fixed?

File: matchers.json (hunk 2 above)
Comment (bryanmacfarlane) [reply matchers.json]: Yeah, I think that is better.  I'll change

File: __tests__/setup-go.test.ts (hunk 3 above)
Comment (ericsciple) [reply __tests__/setup-go.test.ts]: I got confused why a match was returned at all.

The regexp doesn't have `^$` around the pattern. Should it?

Comment (bryanmacfarlane) [reply __tests__/setup-go.test.ts]: I think it matches??  In this line:
```
x./assert.go:5:2
```
it's validating that it only matches the leading dot(s) and not the x.  Not sure this output would ever happen but my initial regex had a bug where I wasn't escaping the dot so it was match 0-2 of any char in from of the slash which is not correct.

Do you have a better suggestion for the title?

File: matchers.json (hunk 2 above)
Comment (bryanmacfarlane) [reply matchers.json]: I changed it to this.

File: __tests__/setup-go.test.ts (hunk 3 above)
Comment (bryanmacfarlane) [reply __tests__/setup-go.test.ts]: fixed.

**Pull request merged**